File: python/thunder/sigprocessing/stats.py
import os
import argparse
import glob
from thunder.sigprocessing.util import SigProcessingMethod
from thunder.util.load import load
from thunder.util.save import save
from pyspark import SparkContext


def stats(data, statistic):
    """compute summary statistics on every data point

    arguments:
    data - RDD of data points
    mode - which statistic to compute ("median", "mean", "std", "norm")

    returns:
    vals - RDD of statistics
    """

    method = SigProcessingMethod.load("stats", statistic=statistic)
    vals = method.calc(data)

    return vals

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="compute summary statistics on time series data")
    parser.add_argument("master", type=str)
    parser.add_argument("datafile", type=str)
    parser.add_argument("outputdir", type=str)
    parser.add_argument("mode", choices=("mean", "median", "std", "norm"),
                        help="which summary statistic")
    parser.add_argument("--preprocess", choices=("raw", "dff", "dff-highpass", "sub"), default="raw", required=False)

    args = parser.parse_args()
    egg = glob.glob(os.path.join(os.environ['THUNDER_EGG'], "*.egg"))
    sc = SparkContext(args.master, "ref", pyFiles=egg)

    data = load(sc, args.datafile, args.preprocess).cache()

    vals = stats(data, args.mode)

    outputdir = args.outputdir + "-stats",

    outputdir = args.outputdir + "-stats"
    if not os.path.exists(outputdir):
        os.makedirs(outputdir)

    save(vals, outputdir, "stats_" + args.mode, "matlab")

File: scala/src/main/scala/thunder/sigprocessing/Stats.scala
package thunder.sigprocessing

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.util.Vector

import thunder.util.Load
import thunder.util.Load._


object Stats {

    def main(args: Array[String]) {
      if (args.length != 5) {
        println("Usage: Stats <master> <datafile> <outputdir> <mode>")
        System.exit(1)
      }
      val sc = new SparkContext(args(0), "Stats")
      val data = Load.loadData(sc, args(1))

      val dims = getDims(data)

      print(Vector(dims.map(_.toDouble)))

      val mode = args(3)
      if (mode == "mean") {
        data.mapValues(x => x.sum / x.length).keys.map(x => x(0)).foreach(println)
      }

    }

}
